### ISSUE TITLE
How to process variable length inputs?

### ISSUE THREAD

**Issue by tlasmanbotch** [HazyResearch/state-spaces#86]
Hello! 

The S4 code as it stands allows the masking of padded inputs based on a given length, but this requires all inputs within a batch to be masked to the same length. Is there an alternative way to deal with padding of different sizes within a given batch (e.g., similar to a self-attention mask)? Would setting masked values to extremely small numbers (e.g. 1e-9) be a viable alternative?

For example, consider the following batch and mask:

```
batch = torch.LongTensor([
    [1, 2, 3, 0, 0, 0],
    [1, 2, 3, 4, 5, 6],
    [1, 2, 3, 4, 5, 0]
])

mask = torch.LongTensor([
    [1, 1, 1, 0, 0, 0],
    [1, 1, 1, 1, 1, 1],
    [1, 1, 1, 1, 1, 0]
])
```

The current code for masking lengths (pasted below) appears to assume all require the same length masking:

```
def forward(self, u, state=None, rate=1.0, lengths=None, **kwargs): # absorbs return_output and transformer src mask
    """
    u: (B H L) if self.transposed else (B L H)
    state: (H N) never needed unless you know what you're doing

    Returns: same shape as u
    """
    if not self.transposed: u = u.transpose(-1, -2)
    L = u.size(-1)

    # Mask out padding tokens
    # TODO handle option for mask - instead of lengths, which assumes suffix padding
    if isinstance(lengths, int):
        if lengths != L:
            lengths = torch.tensor(lengths, dtype=torch.long, device=u.device)
        else:
            lengths = None
    if lengths is not None:
        assert isinstance(lengths, torch.Tensor) and lengths.ndim == 1 and lengths.size(0) in [1, u.size(0)]
        mask = torch.where(torch.arange(L, device=lengths.device) < lengths[:, None, None], 1., 0.)
        u = u * mask
```

From reading through other posts here it seems like S4D is better able to deal with inputs of different lengths than S4, but the S4D code does not contain a masking strategy. 

Thanks in advance for your help!

**Comment by albertfgu** [comment]
The second part of that code snippet assumes `lengths` is a tensor that can be different for every example and creates a mask that plays the same role as your `mask`. Is this not what you want?

**Comment by tlasmanbotch** [comment]
The second part has an assert statement that assumes that the tensor `lengths` only has one dimension. In the example above, `mask` has two-dimensions: a mask for each item in the batch.

If this seems okay, then I can remove the assert statement and provide a mask for each item (instead of lengths of the batch).

**Comment by albertfgu** [comment]
The `mask` in the snippet plays the same role as your mask. There are many ways to modify it to handle the same interface. For example, you can simply comment out the `lengths` logic and just keep the line `u = u * mask` after directly passing in your mask. Or if you don't want to modify the S4 code, you can convert your `mask` into a `lengths` tensor easily (e.g. `lengths = torch.sum(mask, axis=-1)`). Or you can bypass all the masking logic in the S4 module and just do `u = u * mask` outside the module.

**Comment by tlasmanbotch** [comment]
Thanks, just making sure that the logic would not mess with the implementation. Closing issue.
